File: api/agent.ts
import type { VercelRequest, VercelResponse } from './vercel-shim';

type GeminiResponse = {
  candidates?: Array<{
    content?: { parts?: Array<{ text?: string }> };
  }>;
  error?: { message?: string };
};

async function callGemini(prompt: string): Promise<string> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey || apiKey.startsWith('your_')) {
    throw new Error('GEMINI_API_KEY is not configured in Vercel Environment Variables');
  }

  const model = process.env.LLM_MODEL || 'gemini-2.5-flash';
  const url =
    `https://generativelanguage.googleapis.com/v1beta/models/` +
    `${encodeURIComponent(model)}:generateContent?key=${encodeURIComponent(apiKey)}`;

  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      contents: [{ role: 'user', parts: [{ text: prompt }] }],
    }),
  });

  const data = (await response.json()) as GeminiResponse;
  if (!response.ok) {
    throw new Error(data.error?.message || `Gemini HTTP ${response.status}`);
  }

  const text = data.candidates?.[0]?.content?.parts?.map((p) => p.text || '').join('') || '';
  if (!text) {
    throw new Error('Gemini returned an empty response');
  }
  return text;
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ success: false, detail: 'Method not allowed' });
  }

  try {
    const prompt = typeof req.body?.prompt === 'string' ? req.body.prompt.trim() : '';
    if (!prompt) {
      return res.status(400).json({ success: false, detail: 'Missing prompt' });
    }

    const finalOutput = await callGemini(prompt);
    return res.status(200).json({
      success: true,
      pipeline_history: [
        prompt,
        `[Serverless Agent / ${process.env.LLM_MODEL || 'gemini-2.5-flash'}]`,
        finalOutput,
      ],
      final_output: finalOutput,
    });
  } catch (error: unknown) {
    const message = error instanceof Error ? error.message : String(error);
    return res.status(500).json({ success: false, detail: message });
  }
}

File: api/agent.py
from http.server import BaseHTTPRequestHandler
import json
import os
import traceback

import google.generativeai as genai

genai.configure(api_key=os.environ.get("GEMINI_API_KEY", ""))
MODEL_NAME = os.environ.get("LLM_MODEL", "gemini-2.5-flash")


def _cors_headers(handler: BaseHTTPRequestHandler) -> None:
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
    handler.send_header("Access-Control-Allow-Headers", "Content-Type")


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        _cors_headers(self)
        self.end_headers()

    def do_POST(self):
        try:
            content_length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(content_length) if content_length else b"{}"
            payload = json.loads(raw.decode("utf-8") or "{}")
            prompt = (payload.get("prompt") or "").strip()

            if not prompt:
                self._json(400, {"success": False, "detail": "Missing prompt"})
                return

            api_key = os.environ.get("GEMINI_API_KEY", "")
            if not api_key or api_key.startswith("your_"):
                self._json(
                    500,
                    {
                        "success": False,
                        "detail": "GEMINI_API_KEY is not configured in Vercel env",
                    },
                )
                return

            model = genai.GenerativeModel(MODEL_NAME)
            result = model.generate_content(prompt)
            text = getattr(result, "text", None) or str(result)

            self._json(
                200,
                {
                    "success": True,
                    "pipeline_history": [
                        prompt,
                        f"[Serverless Agent Core / {MODEL_NAME}]",
                        text,
                    ],
                    "final_output": text,
                },
            )
        except Exception as exc:
            self._json(
                500,
                {
                    "success": False,
                    "detail": str(exc),
                    "trace": traceback.format_exc(),
                },
            )

    def _json(self, status: int, body: dict) -> None:
        encoded = json.dumps(body).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        _cors_headers(self)
        self.send_header("Content-Length", str(len(encoded)))
        self.end_headers()
        self.wfile.write(encoded)

    def log_message(self, format: str, *args) -> None:
        return
